git bug new

Tatiotaeryy opened the pull request.

Tatiotaeryy committed:
Shitty wizardry

still print warning in windows
what we do is forbidden by all Christian canons

Tatiotaeryy committed:
Now support commands

added support to commands from src/commands

teabagk7 commented:
![](https://github.trello.services/images/mini-trello-icon.png) [Реализовать команду создания нового тикета ('git bug new')](https://trello.com/c/a0646ZAg/2-git-bug-new)

akhismat commented on src/commands/git-bug-new:
```diff
@@ -0,0 +1,62 @@
+#!/bin/sh
+FILE=".viin"
```
Зачем нужен этот файл и почему у него такое имя?

akhismat commented on src/commands/git-bug-new:
```diff
@@ -0,0 +1,62 @@
+	git update-index --add --cacheinfo 100644 $SUMMARY summary
+	git update-index --add --cacheinfo 100644 $DESCRIPTION description
+	TREE=`git write-tree`
+	git rm status >> /dev/null
```
Что такое dev/null? Зачем мы туда что-то пишем?

kevromster replied:
Обычно туда сбрасывают мусорный вывод, который стразу уничтожается и не выводится на экран либо куда-либо ещё.
Видимо, здесь мы не хотим, чтобы пользователь, запустивший этот скрипт, видел логи этих промежуточных команд.
Только обычно с `/dev/null` работают не через `>>`, а через `>`, ну и поток ошибок `stderr` тоже можно спрятать (если надо) вот так:
`echo 'blablabla' > /dev/null 2>&1`

kevromster commented on src/commands/git-bug-new:
```diff
@@ -0,0 +1,62 @@
+
+if [ "$1" = "-m" ]
+then
+	echo $2 > $FILE
```
а если второго параметра нет? нужна проверка корректности входных параметров

kevromster commented on src/commands/git-bug-new:
```diff
@@ -0,0 +1,62 @@
+
+	if [ -f ./.git/index ]
+	then
+		mv ./.git/index ./.git/index2
```
во-первых, почему без `-f` параметра (чтобы не спрашивать у пользователя подтверждения), во-вторых, имя `index2` не очень (а вдруг оно занято будет в новых версиях гита?) - обычно для временных промежуточных файлов лучше использовать в качестве суффикса специальную переменную `$$`, означающую PID текущего процесса.

kevromster commented on src/commands/git-bug-new:
```diff
@@ -0,0 +1,62 @@
+
+	git update-ref $REF $ISSUE
+
+	rm ./.git/index
```
везде `rm -f` лучше использовать, опять же чтобы не спрашивать пользователя подтверждения операции

kevromster commented on src/commands/git-bug-new:
```diff
@@ -0,0 +1,62 @@
+
+fi
+
+if [ -f $FILE ]
```
Эти четыре строчки if'а можно переписать в одну так: `[ <условие> ] && <действие>` - не только здесь, но и в других аналогичных местах этого скрипта.
А конкретно тут так, с кавычками вокруг `$FILE`, так как в его имени или пути могут быть пробелы:
`[ -f "$FILE" ] && rm -f "$FILE"`

Tatiotaeryy committed:
Minor fix

- make install now works correctly

Tatiotaeryy committed:
Major fix

- ref "ISSUES" now stored in refs/heads/
- now it works in accordance with the model in MODEl.md
- no longer displays the error in Windows

kevromster commented on Makefile:
```diff
@@ -14,10 +14,12 @@ all:
 install:
 	mkdir -p $(SRC_PATH)
 	if [ -f $(SRC_PATH)/$(SRC_OBJ) ] ; then  rm $(SRC_PATH)/$(SRC_OBJ) ; fi;
+	find "$(SRC_PATH)" -name '$(SRC_OBJ)-*' -exec rm {} \;
```
Непонятно, почему фиксы Makefile делаются в этой ветке - они же не относятся к задаче?
Если нашлась какая-то ошибка в коде, не связанном с производимым изменением, заводится отдельная бага в трэкере задач и фиксится в своей ветке.
В данном случае пусть @AnnaKhismatullina решит, выносить это в отдельную ветку или оставить уже тут на первый раз :)

kevromster commented on src/commands/git-bug-new:
```diff
@@ -0,0 +1,51 @@
+#!/bin/bash
+
+FILE=".viin"
```
уже спрашивалось, почему такое страное имя? лучше какой-нибудь `/tmp/blabla$$` создать, а то здесь рискуем пересечься с файлами пользователя репозитория.

kevromster commented on src/commands/git-bug-new:
```diff
@@ -0,0 +1,51 @@
+	"${EDITOR:-vi}" $FILE
+fi
+
+if [ -s "$FILE" ]
```
предлагаю инвертировать условие - если файла нет, то сразу exit, можно с выводом в консоль, что не предоставлено описание ишуя

kevromster commented on src/commands/git-bug-new:
```diff
@@ -0,0 +1,51 @@
+	[ -f "./.git/index" ] && mv -f "./.git/index" "./.git/index$$"
+
+	STATUS=`echo open | git hash-object -w --stdin`
+	LINE=`head -1 $FILE`
```
А вдруг юзер пустой файл оставил или там одними пробелами заполнил? Кажется, перед любыми манипуляциями с `git`, стоит проверить, что контент не пуст. Ну и LINE стоит strip'нуть от окружающих пробелов. Плюс поставить тут и в других местах двойные кавычки.

Tatiotaeryy replied:
Спасибо, исправили.

kevromster commented on src/commands/git-bug-new:
```diff
@@ -0,0 +1,51 @@
+	SUMMARY=`echo $LINE | git hash-object -w --stdin`
+	DESCRIPTION=`sed -i 1d $FILE | git hash-object -w $FILE 2>/dev/null`
+
+	git update-index --add --cacheinfo 100644 $STATUS status
```
кажется, здесь и ниже нужно в кавычках писать "$STATUS", "$DESCRIPTION", etc? чтобы длинные строки с пробелами правильно обрабатывались

Tatiotaeryy replied:
Здесь в STATUS, DESCRIPTION, SUMMARY хранятся хэши соответствующих блобов, насколько мне известно в шестнадцатеричную запись пробелы не входят.

kevromster commented on src/commands/git-bug-new:
```diff
@@ -0,0 +1,51 @@
+
+if [ -s "$FILE" ]
+then
+	[ -f "./.git/index" ] && mv -f "./.git/index" "./.git/index$$"
```
в другом PR уже писал - зависим от относительного местоположения и думаем, что запускаем скрипт из корня проекта. А если я в подпапку перейду?

kevromster commented on src/commands/git-bug-new:
```diff
@@ -0,0 +1,51 @@
+	fi
+	git update-ref $REF $ISSUE
+
+	rm -f "./.git/index"
```
тот же вопрос, что и в другом PR - зачем этот файл тут удалять? следующий же `mv` его просто перетрёт

Tatiotaeryy replied:
Если перемещаемого файла нет, то затирать его будет нечем, а следы подчищать надо.

Tatiotaeryy committed:
LACRIMOSA

- now we definitely look to .git directory
- tried to make it more self-documenting

kevromster commented on src/commands/git-bug-new:
```diff
@@ -0,0 +1,66 @@
+
+get_git_dir()
+{
+	until [ -d ./.git ]; do cd ..; done; echo "$PWD/.git";
```
А если _.git_ так и не найдётся?

kevromster commented on src/commands/git-bug-new:
```diff
@@ -0,0 +1,66 @@
+	REF="refs/heads/ISSUES"
+
+	local GIT_DIR="`get_git_dir`"
+	local FILE="$GIT_DIR/ISSUE_EDIT"
```
1) Может, добавить `$$` к имени временного файла? Если в текущем виде ты два раза одновременно запустишь скрипт в одном репе, что произойдёт?
2) Непонятно, зачем его в $GIT_DIR прятать. Если вдруг скрипт развалится посередине, пользователь его и не увидит.

Tatiotaeryy replied:
1) Ну собсна он выдаст алярму что такой файл уже открыт и редактируется, это же например произойдёт если сумасшедший юзер попробует одновременно создать два коммита, не совсем понимаю для чего такие зверства, но если скажете что это необходимо сделаю.
2) Чтоб уж точно не пересечься с файлами сумасшедшего пользователя дающего файлам имена начинающееся с точки.

kevromster commented on src/commands/git-bug-new:
```diff
@@ -0,0 +1,66 @@
+	fi
+
+	[ ! -f "$FILE" ] && exit 0
+	[ ! -s "$FILE" ] && rm -f "$FILE" && exit 0
```
ну, строго говоря, тут всегда нужно делать `exit`, даже если `rm` почему-то вернёт ошибку.
Касательно всего кода: обычно, если выходим с ошибкой, то это не 0, а, например, 1.

kevromster commented on src/commands/git-bug-new:
```diff
@@ -0,0 +1,66 @@
+
+	local LINE="$(echo $(sed -n 1p $FILE | awk '$1 != "" {print $0}'))"
+
+	[ "$LINE" = "" ] && rm -f "$FILE" && exit 0
```
аналогично с `&& exit 0`

Tatiotaeryy committed:
δεκανίκι

- removed endless loop when there not a git repository
- if we can't remove file, we exit anyway

kevromster commented on src/commands/git-bug-new:
```diff
@@ -0,0 +1,77 @@
+	local GIT_DIR="$(get_git_dir)"
+	if [ "$GIT_DIR" = "" ]
+	then
+		echo "Error: There's nothing here"
```
пользователь ничего не поймёт, увидев такое сообщение - нужно написать, чего именно не нашлось

kevromster commented on src/commands/git-bug-new:
```diff
@@ -0,0 +1,77 @@
+		exit 1
+	fi
+	
+	local FILE="$GIT_DIR/ISSUE_EDIT"
```
Продолжу дискуссию по этой строчке - `$$` спасёт отца русской демократии от  совпадения с файлами пользователя, и внутри $GIT_DIR держать этот файл нет смысла.

Tatiotaeryy replied:
Если мы полагаем пользователя настолько сумасбродным кто запрещает ему добавить в конце имени файла случайный набор цифр? Баба яга против!

kevromster replied:
никто же не мешает проверить, что такого файла нет у пользователя, а если есть - добавить единичку, например

Tatiotaeryy replied:
А, собственно, зачем делать лишнюю проверку если можно просто положить туда куда ему лезть не стоит?
Мне кажется так лучше.

kevromster commented on src/commands/git-bug-new:
```diff
@@ -0,0 +1,77 @@
+{
+	local dir="$PWD"
+	until [ -d ./.git ]; do cd ..; if [ "$PWD" = "$dir" ]; then exit 1; fi; local dir="$PWD"; done; echo "$PWD/.git";
+	#Не совсем уверен как проверять невозможность поднятся выше, поэтому такой вот костыль
```
вообще, первая же гугл-ссылка даёт ответ :)
`git rev-parse --show-toplevel`

Tatiotaeryy replied:
Полезный совет, спасибо.

kevromster commented on src/commands/git-bug-new:
```diff
@@ -0,0 +1,77 @@
+
+	[ ! -f "$FILE" ] && exit 0
+
+	local LINE="$(echo $(sed -n 1p $FILE | awk '$1 != "" {print $0}'))"
```
а что этот код делает? первую непустую строчку из файла получает?

kevromster commented on src/commands/git-bug-new:
```diff
@@ -0,0 +1,77 @@
+	local DESCRIPTION=$(sed -i 1d $FILE | git hash-object -w $FILE 2>/dev/null)
+
+	git update-index --add --cacheinfo 100644 $STATUS status
+	git update-index --add --cacheinfo 100644 $SUMMARY summary
```
в кавычки `$SUMMARY` и `$DESCRIPTION` (`$STATUS` заодно тоже можно)?

Tatiotaeryy replied:
nope

kevromster replied:
хочу развёрнутый ответ почему нет

Tatiotaeryy replied:
Хм, вроде бы я уже писал, ну да ладно:
В этих переменных хранятся хэши соответствующих блобов, так как шестнадцатеричное число не содержит в своей записи пробелы это не нужно.

kevromster replied:
Рекомендую пересмотреть лекцию о качестве кода. Конкретно здесь проблема - плохие названия переменных - в `$SUMMARY` хранится не саммари, а хэш, и т.д.

Tatiotaeryy replied:
Чтож. Изменю названия этих переменных, и подработаю прочее, что касается читаемости и самодокументированности, но в данном конкретном случае претензии мне всё же кажутся надуманными. Это как с названиями файлов никто не даёт файлу имя "Имя файла хранящего такое-то содержимое", все просто называют его "Такое-то содержимое"; хэш для гита выступает в качестве имени файла, поэтому такой вариант мне казался очевидным.

kevromster commented on src/commands/git-bug-new:
```diff
@@ -0,0 +1,77 @@
+	echo $ISSUE	
+
+	local HASH=`echo $ISSUE | git hash-object -w --stdin`
+	if git show-ref --quiet --verify $REF
```
почему бы комментарий на добавить о том, что именно этот `if` здесь проверяет?

Tatiotaeryy committed:
Meaningless update

- now we search git repository with git commands
- added some comments

kevromster commented on src/commands/git-bug-new:
```diff
@@ -0,0 +1,71 @@
+	REF="refs/heads/ISSUES"
+
+	local GIT_DIR="$(git rev-parse --show-toplevel)"/.git
+	if [ "$GIT_DIR" = "" ]
```
Он пустой никогда не будет сейчас. Лучше проверить на -d

teabagk7 committed:
8

-changed names of variables
-added comments

kevromster commented on src/commands/git-bug-new:
```diff
@@ -0,0 +1,77 @@
+
+#ищем путь до репозитория и проверяем на его наличие
+	local GIT_DIR="$(git rev-parse --show-toplevel)"/.git
+	if [ -d "$GIT_DIR" ]
```
Инвертировать условие нужно.

Tatiotaeryy replied:
Это одна из ошибок, извините.)

kevromster commented on src/commands/git-bug-new:
```diff
@@ -0,0 +1,77 @@
+	local GIT_DIR="$(git rev-parse --show-toplevel)"/.git
+	if [ -d "$GIT_DIR" ]
+	then
+		exit 1
```
тихий выход для пользователя - непонятно, нужно написать что-нибудь. В других местах тоже стоит добавить текстовое сообщение.

Tatiotaeryy replied:
В этом случае будет выведено стандартное сообщение гита об отсутствии репозитория, вызванное 8 строкой.

kevromster commented on src/commands/git-bug-new:
```diff
@@ -0,0 +1,77 @@
+	then
+		echo "$2" > $FILE
+	else
+		"${EDITOR:-vi}" $FILE
```
У меня при тестировании vi не смог создать/сохранить файл ISSUE_EDIT.
Скрипт вообще тестировался на работоспособность? )

Tatiotaeryy replied:
Нет, мы случайно залили неисправленную версию с ошибками, сейчас всё исправим)

Tatiotaeryy committed:
9

- fixed some critical bugs.

kevromster reviewed (approved):
LGTM
